python/ray/tune/api_v2/tuner.py

```
@@ -0,0 +1,345 @@
+            pickle.dump(self, fp)
+
+        trainable_ckpt = os.path.join(self.experiment_checkpoint_dir, "trainable.pkl")
+        with open(trainable_ckpt, "wb") as fp:
+            pickle.dump(self.trainable, fp)
+
+    def _process_dataset_param(self):
+        """Dataset needs to be fully executed before sent over to trainables.
+
+        A valid dataset configuration in param space looks like:
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        """
+
+        def _helper(dataset_dict: dict):
+            for k, v in dataset_dict.items():
+                if isinstance(v, dict):
+                    _helper(v)
+                elif isinstance(v, Dataset):
+                    dataset_dict[k] = v.fully_executed()
+                # TODO(xwjiang): Consider CV config for beta.
+                # elif isinstance(v, int):
+                #     # CV settings
+                #     pass
+                elif isinstance(v, list):
+                    if not all([isinstance(v_item, int) for v_item in v]) and not all(
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if _DATASETS in self.param_space:
+            ds = self.param_space[_DATASETS]
+            if isinstance(ds, Dataset):
+                self.param_space[_DATASETS] = ds.fully_executed()
+            elif isinstance(ds, dict):
+                _helper(ds)
+            else:
+                # We shouldn't be expecting anything here.
+                raise TuneError("Unexpected dataset param passed in.")
```

Can we raise a better error message here?